### PR TITLE
Routing parameters substitution

### PR DESCRIPTION
**Pull request by matej21**
### parameters subsitution
RoutingParameter class as a substitution placeholder can be used in `parameters` in RoutingLinkTarget.
- In every scope, parameters from request are accessible through `request` prefix, e.g. `request.id`
- In data binding context, you can reference any scalar field (relative single field) using `entity` prefix. e.g `entity.id`
- Additional parameters passed through `parameters` prop are accessible without a prefix

## links DSL
String format of a link target now support parameters with following syntax:
```
pageName(param: value[... , param: value])
```
value can be either a scalar (number or single quoted string) or a variable variable starts with a dollar `$` and is followed by parameter identifier (see parameter substitution)

example:
```tsx
<PageLink to="articleEdit(id: $entity.id, anotherParam: $paramValue)" parameters={{paramValue: foo}}>Edit</PageLink>
```

### DSL limitation:
- Nested parameters are not supported

## Backward incomplatible changes

- PageConfig object with an alternative request representation has been removed

## Target release

This PR targets version 0.8.

## Other considerations

PageLinkById component is now obsolete, so we should consider deprecation or removal of this component.

### CODE REVIEW

File: packages/admin/src/routing/useRoutingLink.ts [hunk 1]
```diff
@@ -28,20 +35,52 @@ const targetToRequest = (target: RoutingLinkTarget, currentRequest: RequestState
...
+			return [name, resolveParameter(value.name)]
+		}
+		if (typeof value === 'object' && value !== null) {
+			return [name, resolveParameters(value, resolveParameter)]
+		}
+		return [name, value]
+	}))
+}
+
+const completeRequest = (request: IncompleteRequestState, currentRequest: RequestState, resolveParameter: RoutingParameterResolver): RequestState => {
```
Comment (VojtaStanek) [inline packages/admin/src/routing/useRoutingLink.ts:53]: I would inline this function

File: packages/admin/src/routing/types.ts [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
 	parameters: P
 	dimensions: SelectedDimension
 }
 
-export type RequestState = PageRequest<any> | null
+export type RequestState<Parameters extends RequestParameters<RoutingParameter> = RequestParameters> = PageRequest<Parameters> | null
+
+export type RequestChange = (currentState: RequestState) => IncompleteRequestState
+export type DynamicRequestParameters = RequestParameters<RoutingParameter>;
+export type IncompleteRequestState = Partial<RequestState<DynamicRequestParameters>> & { pageName: string } | null
```
Comment (VojtaStanek) [inline packages/admin/src/routing/types.ts:23]: Why union with null?

File: packages/admin/src/routing/types.ts [hunk 3]
```diff
@@ -1,20 +1,29 @@
+import { RoutingParameter } from './RoutingParameter'
+
 export interface SelectedDimension {
 	[key: string]: string[]
 }
 
-export interface PageParameters {
-	[key: string]: string | PageParameters | undefined
+export type RequestParameterValue<Extra extends RoutingParameter = never> = string | undefined | Extra;
```
Comment (VojtaStanek) [inline packages/admin/src/routing/types.ts:7]: Why is it generic? Wouldn't it be better (more readable) to have the extra union on "use-side"?

File: packages/admin/src/routing/useRoutingLink.ts [hunk 4]
```diff
@@ -28,20 +35,52 @@ const targetToRequest = (target: RoutingLinkTarget, currentRequest: RequestState
...
+	return target
+}
+
+const resolveParameters = (parameters: DynamicRequestParameters, resolveParameter: RoutingParameterResolver): RequestParameters => {
+	return Object.fromEntries(Object.entries(parameters).map(([name, value]) => {
+		if (value instanceof RoutingParameter) {
+			return [name, resolveParameter(value.name)]
+		}
+		if (typeof value === 'object' && value !== null) {
+			return [name, resolveParameters(value, resolveParameter)]
```
Comment (VojtaStanek) [inline packages/admin/src/routing/useRoutingLink.ts:47]: Why this recursive call?

**Review by VojtaStanek — CHANGES_REQUESTED** [review]

File: packages/binding/src/queryLanguage/Parser.ts [hunk 5]
```diff
@@ -771,6 +839,7 @@ namespace Parser {
 		uniqueWhere: UniqueWhere // E.g. (author.mother.id = 123)
 		filter: Filter // E.g. [author.son.age < 123]
 		orderBy: OrderBy // E.g. items.order asc, items.content.name asc
+		functionCall: ParsedFunctionCall
```
Comment (JanTvrdik) [inline packages/binding/src/queryLanguage/Parser.ts:842]: e.g.

**Pull request merged**